rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,364 @@
+
+Alternatives:
+- Use `unit` in the root directory instead
+  - (+) This is future proof in case we want to make the directory structure more general purpose
+    - (-) We don't yet know if we want that, so this is out of scope for now
+- Use `pkgs` instead, so that the `${shard}`'s are siblings to the other current directories in `pkgs` such as `top-level`, with the intention that the other directories would be hopefully removed at some point, then only leaving the shards in `pkgs`
+  - (+) If we remove the other directories at some point, only the `${shard}`'s will be left in `pkgs`
+  - (-) This leads to ambiguities between the directories from the standard and the other directories, requiring special handling in the code and CI, leading to complexities.
+  - (-) This would be harder to document and explain to people, since one always has to disregard all non-unit directories, with no obvious justification
+  - (-) Currently we cannot apply this standard to all definitions in `pkgs`, in particular nested packages like `pythonPackages.*`, non-`callPackage`'d definitions like `copyDesktopItems` and non-derivations like `fetchFromGitHub`.
+    Depending on how we want to handle those, it might make more sense to keep `pkgs/unit` or to use `pkgs` directly once all legacy paths are migrated away to another top-level directory, we don't yet know. `pkgs/unit` will be easier to migrate to `pkgs` than the other way around though.
+- Another name than `unit`, proposed were `auto`, `pkg`, `mod`, `component`, `part`, `comp`, `app`, `simple`
+  - (-) `unit` is considered the most appropriate because a "unit" typically refers to a discrete, indivisible entity that is distict from other entities of the same type.
+  - (-) In addition we envision that in the future we would extend the unit directory standard to not just include a package definition for each unit, but also other parts such as NixOS modules, library components, tests, etc. In this case `unit` would fit even better and could be described as
+    > A collection of standardized files related to the same software component
+
+
+## Alternate shard structure
+
+Context: The structure is `pkgs/unit/${shard}/${name}` with `shard` being the lowercased two-letter prefix of `name`.
+
+Alternatives:
+- A flat directory, where `pkgs.hello` would be in `pkgs/unit/hello`.
+  - (+) Simpler for the user and code.
+  - (-) The GitHub web interface only renders the first 1 000 entries when browsing directories, which would make most packages inaccessible in this way.
+    - (+) This feature is not used often.
+      - (-) [A poll](https://discourse.nixos.org/t/poll-how-often-do-you-navigate-through-directories-on-github-to-get-to-a-nixpkgs-package-file/21641) showed that about 41% of people rely on this feature every week.
+  - (-) Bad because it makes `git` and file listings slower.
+- Use three-letter or four-letter prefixes.
+  - (-) Also leads to directories containing more than 1 000 entries, see above.
```

Comment:
Anyone considered using, say, the first and last character of the name instead of a prefix? I tried that and the largest shard had 202 entries (though I was lazy and just checked all the attr names of the top-level object, instead of doing whatever is the right thing for listing all packages and only packages). Still only two characters and one level, still fully determined by the name alone.

Reply:
Interesting idea! By adjusting the [gist to calculate the distribution of shards](https://gist.github.com/infinisil/95c7013db62e9f23ab2bc92165a37221) to this:

<details>

```
{
  system ? builtins.currentSystem,
  nixpkgs ? fetchTarball {
    url = "https://github.com/NixOS/nixpkgs/tarball/6948ef4deff7a72ebe5242244bd3730e8542b925";
    sha256 = "1fw8vlbcxr3kw96rrjlw94zhbjs3a32dcnvav19cvniwlbhf2jsh";
  },
}:
let
  pkgs = import nixpkgs {
    inherit system;
    config.allowAliases = false;
    overlays = [];
  };
  inherit (pkgs) lib;

  set = pkgs;

  names = lib.attrNames set;
  attributeCount = lib.length names;

  grouped = lib.groupBy (name:
    lib.toLower (lib.substring 0 1 name + lib.substring (lib.stringLength name - 1) 1 name)
  ) names;

  distribution = lib.sort (a: b: a.count < b.count)
    (lib.mapAttrsToList (name: values: {
      inherit name;
      count = lib.length values;
    }) grouped);

  shardCount = lib.length distribution;

  top20 = lib.drop (shardCount - 20) distribution;

  result =
    builtins.trace "Attribute count: ${toString attributeCount}"
    builtins.trace "Shard count: ${toString shardCount}"
    builtins.trace "Top 20"
    (lib.foldl' (acc: el:
      builtins.trace "${el.name}: ${toString el.count}" acc
    ) null top20);

in {
  inherit result;
}
```
</details>

I'm getting this distribution:

```
$ nix-instantiate shards.nix --eval -A result
trace: Attribute count: 17305
trace: Shard count: 895
trace: Top 20
trace: lr: 104
trace: pt: 104
trace: gt: 106
trace: as: 108
trace: mr: 110
trace: se: 110
trace: pe: 113
trace: ms: 114
trace: ce: 120
trace: lt: 129
trace: ge: 134
trace: st: 134
trace: ls: 135
trace: cs: 136
trace: sr: 137
trace: le: 138
trace: gs: 155
trace: ss: 160
trace: pr: 171
trace: ps: 183
```

For comparison, here's the one with the 2-letter prefix:

<details>

```
trace: Shard count: 726
trace: Top 20
trace: gr: 121
trace: do: 133
trace: mi: 133
trace: gn: 137
trace: ne: 139
trace: po: 143
trace: st: 144
trace: mo: 149
trace: cl: 161
trace: re: 166
trace: gi: 172
trace: pa: 184
trace: te: 184
trace: go: 210
trace: pr: 217
trace: ca: 243
trace: ma: 244
trace: co: 252
trace: op: 260
trace: li: 1092
```
</details>

And simulating a larger package count using
```nix
pkgs // pkgs.haskellPackages // pkgs.python3Packages // pkgs.perlPackages // pkgs.emacsPackages;
```

```
trace: Attribute count: 47078
trace: Shard count: 951
trace: Top 20
trace: sr: 348
trace: de: 351
trace: ae: 357
trace: st: 361
trace: ts: 366
trace: me: 367
trace: ms: 367
trace: he: 368
trace: gs: 369
trace: ge: 390
trace: hs: 401
trace: pr: 403
trace: pt: 412
trace: as: 423
trace: cs: 433
trace: ce: 489
trace: se: 551
trace: ss: 571
trace: ps: 574
trace: pe: 583
null
```

And the 2-letter prefix for comparison:
<details>

```
trace: Attribute count: 47078
trace: Shard count: 804
trace: Top 20
trace: da: 405
trace: he: 413
trace: op: 436
trace: fl: 441
trace: po: 474
trace: se: 479
trace: di: 487
trace: go: 572
trace: st: 573
trace: pr: 587
trace: pa: 603
trace: mo: 606
trace: ma: 610
trace: ca: 628
trace: te: 685
trace: ha: 733
trace: re: 738
trace: co: 1031
trace: py: 1194
trace: li: 1423
```
</details>

It's really nice how the distribution is much more equal, keeping the number of directories in shards low even with many more packages!

Though I also see a number of disadvantages:
- It's not as obvious what the sharding scheme is. 2-letter prefix is slightly more obvious from the paths themselves (`pkgs/unit/he/hello`), and much more obvious when seeing all packages in the same directory having the same prefix (`pkgs/unit/he` only contains entries starting with `he`)
- Similarly-named packages won't be close to each other anymore, `hello` will be in `ho`, but `hello-unfree` in `he`.
- The directory listing for `pkgs/unit` won't be attribute-alphabetical
- Changing the suffix of attributes (or creating new ones with a different suffix) requires moving them to a different shard.

Reply:
I find the approach really interesting, but for me the major downside is the added cognitive overhead when navigating — even if I know the scheme. Try to pick a random package name (preferably without reading its name on screen so that you have to do the calculation in your head) and type its first two letters vs its first and last.

But this got me thinking: seeing both lists @infinisil generated, to me they look pretty similar except for `li` because of `lib`. So how about we hard-code special-case some exceptions, so that packages that start with `lib` get a five letter shard? (Alternatively, strip lib from the name before sharding so `libfoo` would land in `fo` instead of `li`)

Reply:
I don't think it's a big problem that the `li` shard has over 1000 entries. Yeah GitHub doesn't display all of them in the directory view, but that's very minor (only about 100 packages aren't reachable directly from the file browser!). In comparison, any exception would add a non-trivial mental overhead and potential confusion. So I don't think we should do that.

Reply:
It's always a tradeoff, and I have no strong opinions here. We should document these idea of the Alternatives section though.

Reply:
i think doing an exception for lib is trivial enough that the benefits are greater than the added mental-overhead

Reply:
I'd like to explore the idea of a five-letter shard for "lib" some more, and would like to see the result of the resulting distribution. My main concern here is that in theory it may result in more than 1000 shards in the `unit` directory, although I doubt we'll get close to that in practice anytime soon.

Reply:
if we do `lib.toLower (lib.substring 1 2 name)`, we get `ib` as largest shard with 935, which would fit into the github limit, but that would be ridiculous and not less confusing than using the first and last character
although the grouping of e.g. `hello` next to `hello_unfree` would work

Reply:
I just noticed that already today we have three directories in Nixpkgs with more than 1000 entries!
- [`pkgs/development/libraries`](https://github.com/NixOS/nixpkgs/tree/7d9999f2c178882a4656283c12a7501243626bf7/pkgs/development/libraries) with 1995 entries (my birth year! :birthday:), so probably already about 995 libraries are currently unavailable directly from GitHub's file browser!
- [`pkgs/development/python-modules`](https://github.com/NixOS/nixpkgs/tree/7d9999f2c178882a4656283c12a7501243626bf7/pkgs/development/python-modules) with 6145 entries! This is currently an effectively flat listing of all python packages, so this would be fixed if we used the same sharding in the future.
- [`pkgs/tools/misc`](https://github.com/NixOS/nixpkgs/tree/7d9999f2c178882a4656283c12a7501243626bf7/pkgs/tools/misc) with 1013 entries, an additional 13 entries not directly accessible

And I don't think I've ever seen anybody complain about this, so I don't think we should be worried about this proposal ending up with only ~100 non-directly accessible packages.

Reply:
Although i don't use GitHub to browse the Nixpkgs source code anymore (i navigate to package sources by clicking on `Source` on search.nixos.org), i found the GitHub limitations quite puzzling and frustrating when i first started out with Nix.

Reply:
> Alternatively, strip lib from the name before sharding so `libfoo` would land in `fo` instead of `li`.

I liked this thought; it shouldn't blow up the number of shards, and if putting related packages in the same shard is a goal then any `foo`/`libfoo` pairs we have will end up together.

I wouldn't object to any of the ideas proposed, but like @Quantenzitrone I do find it occasionally frustrating not to be able to make maximum use of GitHub's affordances.

Reply:
A downside of making an exception for lib* packages, are packages, that start with lib but are not a library like e.g. `librarian-puppet-go`.

Reply:
It also opens up the possibility for collisions, e.g. `libfoo` vs `foo`, and not that unlikely ones.

Reply:
no, the `name` will still be `libfoo`, so no collision with `foo` happens, otherwise the autocaller wouldn't that libfoo is a library
e.g.
```
fo
|- foo
|- foobar
'- libfoo
```

Reply:
The flipside to the GitHub limitations being annoying is that, if GitHub being annoying pushes somebody from navigating on GitHub to navigating a local clone, they're one step closer to contributing!

Reply:
if that is the goal, we should use the flat directory structure proposed in the alternatives

Reply:
> I'd like to explore the idea of a five-letter shard for "lib" some more, and would like to see the result of the resulting distribution. My main concern here is that in theory it may result in more than 1000 shards in the `unit` directory, although I doubt we'll get close to that in practice anytime soon.

how about a 4 letter shard for `lib*` packages tho?
there are only 909 packages that start with lib, so it doesn't really make sense to split these into 323 different shards, but when using 4 letter shards for `lib*` packages we only have 29 shards for the lib packages with `libs` being the biggest shard with 70 packages
this results in only 755 total shards and has the benefit of it being easier to understand due to the shard being the start of the package attribute name (just a longer substring for lib*)

Reply:
Another argument for not adding such a special case: We can expect to eventually want to use this standard for other package sets too, such as Python packages. Keeping the same special case as for top-level packages doesn't make sense then. One might argue that we should be able to add arbitrary special cases depending on the package set, but that does introduce another variable to the game, increasing complexity again.

From the arguments brought up so far I'm fully convinced that we shouldn't do this, the benefits are not high enough to justify the cost.

Reply:
yeah for python we would need a special case for the py shard, as it currently has about 1070 packages

Reply:
Thinking about this some more, there's a good case to be made that all these library packages should really be in their own package set, since they're not generally something users would directly install (which is somewhat of a convention with top-level packages). So we might want to have `lib.png` instead of `libpng`.

Reply:
I don't agree with this.  The convention is not about whether something is directly installable, but whether it makes sense for it to be user accessible, and whether it belongs in a separate namespace to avoid collisions (e.g. `pythonPackages`).  Directly using (albeit not installing) a library package is extremely common: `nix-shell -p libpng` when compiling a small program.  Such a change would make Nixpkgs pointlessly different to every other package repository.

Reply:
I'll also note that we don't strip prefixes even for packages in namespaces — look at all the package names starting with "python-" in python-packages.nix.

Reply:
The structure of Nixpkgs doesn't have to be exposed directly to the user though. I could imagine a future in which we expose packages both stripped (for a clean Nixpkgs tree with unit directories) and unstripped (for upstream names). Anyways, this is all just an "could be", not a "have to". I'll make sure to add all the arguments mentioned here to the RFC with that in mind.

Reply:
> I'll also note that we don't strip prefixes even for packages in namespaces — look at all the package names starting with "python-" in python-packages.nix.

actually for some package sets we do, look at fishPlugins for example

Reply:
> Thinking about this some more, there's a good case to be made that all these library packages should really be in their own package set, since they're not generally something users would directly install (which is somewhat of a convention with top-level packages). So we might want to have `lib.png` instead of `libpng`.

How exactly would a library be defined? With package sets like pythonPackages or vimPlugins, this is mostly unambiguous, but for libraries, anything can have a `lib` output (right?).
Furthermore, not all libraries start with lib.

Reply:
i honestly would just leave the unit structure as it is right now, having one shard with over 1000 entries is acceptable and definitely better than the current package structure, so let's leave possible shard exceptions to later rfc's and get this merged

Reply:
> I'll also note that we don't strip prefixes even for packages in namespaces — look at all the package names starting with "python-" in python-packages.nix.

This is generally done on purpose to follow the upstream/PyPI names.

Reply:
Have we discussed single-letter prefix already? Someone argued that the 1000 file folder limit of GitHub is pretty arbitrary as restriction, and while I do think we should try to bound the folder size, I'd like to know how bad it gets with single-letter prefixes.

Reply:
```
trace: Attribute count: 17305
trace: Shard count: 27
trace: Top 20
trace: _: 31
trace: y: 134
trace: q: 210
trace: z: 213
trace: j: 270
trace: v: 311
trace: u: 340
trace: x: 354
trace: k: 399
trace: i: 426
trace: w: 430
trace: h: 440
trace: e: 479
trace: o: 529
trace: n: 634
trace: b: 647
trace: r: 662
trace: f: 691
trace: d: 708
trace: t: 801
trace: a: 854
trace: m: 1039
trace: p: 1217
trace: c: 1246
trace: g: 1268
trace: s: 1424
trace: l: 1548
```